File: gateways/tunnel-agent/src/config.test.ts
import fs from 'fs';
import os from 'os';
import path from 'path';
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { loadConfig } from './config';

describe('loadConfig', () => {
  const ORIGINAL_ENV = { ...process.env };

  beforeEach(() => {
    // Clean tunnel vars before each test
    delete process.env.TUNNEL_SERVER_URL;
    delete process.env.TUNNEL_TOKEN;
    delete process.env.TUNNEL_GATEWAY_ID;
    delete process.env.TUNNEL_LOCAL_PORT;
    delete process.env.TUNNEL_LOCAL_HOST;
    delete process.env.TUNNEL_CA_CERT;
    delete process.env.TUNNEL_CA_CERT_FILE;
    delete process.env.TUNNEL_CLIENT_CERT;
    delete process.env.TUNNEL_CLIENT_CERT_FILE;
    delete process.env.TUNNEL_CLIENT_KEY;
    delete process.env.TUNNEL_CLIENT_KEY_FILE;
    delete process.env.TUNNEL_PING_INTERVAL_MS;
    delete process.env.TUNNEL_RECONNECT_INITIAL_MS;
    delete process.env.TUNNEL_RECONNECT_MAX_MS;
    delete process.env.TUNNEL_AGENT_VERSION;
  });

  afterEach(() => {
    // Restore env — delete any TUNNEL_ vars added during the test
    const tunnelKeys = [
      'TUNNEL_SERVER_URL', 'TUNNEL_TOKEN', 'TUNNEL_GATEWAY_ID', 'TUNNEL_LOCAL_PORT',
      'TUNNEL_LOCAL_HOST', 'TUNNEL_CA_CERT', 'TUNNEL_CA_CERT_FILE', 'TUNNEL_CLIENT_CERT',
      'TUNNEL_CLIENT_CERT_FILE', 'TUNNEL_CLIENT_KEY', 'TUNNEL_CLIENT_KEY_FILE',
      'TUNNEL_PING_INTERVAL_MS', 'TUNNEL_RECONNECT_INITIAL_MS', 'TUNNEL_RECONNECT_MAX_MS',
      'TUNNEL_AGENT_VERSION',
    ] as const;
    for (const key of tunnelKeys) {
      if (!(key in ORIGINAL_ENV)) {
        process.env[key] = undefined;
      }
    }
  });

  it('returns null when no tunnel env vars are set (dormant mode)', () => {
    const cfg = loadConfig();
    expect(cfg).toBeNull();
  });

  it('returns a valid config when all required vars are set', () => {
    process.env.TUNNEL_SERVER_URL = 'wss://example.com/tunnel';
    process.env.TUNNEL_TOKEN = 'abc123';
    process.env.TUNNEL_GATEWAY_ID = 'gw-uuid-001';
    process.env.TUNNEL_LOCAL_PORT = '4822';

    const cfg = loadConfig();
    expect(cfg).not.toBeNull();
    expect(cfg!.serverUrl).toBe('wss://example.com/tunnel');
    expect(cfg!.token).toBe('abc123');
    expect(cfg!.gatewayId).toBe('gw-uuid-001');
    expect(cfg!.localServicePort).toBe(4822);
    expect(cfg!.localServiceHost).toBe('127.0.0.1');
  });

  it('uses default values for optional settings', () => {
    process.env.TUNNEL_SERVER_URL = 'wss://example.com/tunnel';
    process.env.TUNNEL_TOKEN = 'abc123';
    process.env.TUNNEL_GATEWAY_ID = 'gw-uuid-001';
    process.env.TUNNEL_LOCAL_PORT = '2222';

    const cfg = loadConfig();
    expect(cfg!.pingIntervalMs).toBe(15000);
    expect(cfg!.reconnectInitialMs).toBe(1000);
    expect(cfg!.reconnectMaxMs).toBe(60000);
    expect(cfg!.localServiceHost).toBe('127.0.0.1');
    expect(cfg!.caCert).toBeUndefined();
    expect(cfg!.clientCert).toBeUndefined();
    expect(cfg!.clientKey).toBeUndefined();
  });

  it('includes optional TLS values when set', () => {
    process.env.TUNNEL_SERVER_URL = 'wss://example.com/tunnel';
    process.env.TUNNEL_TOKEN = 'tok';
    process.env.TUNNEL_GATEWAY_ID = 'gw-1';
    process.env.TUNNEL_LOCAL_PORT = '4822';
    process.env.TUNNEL_CA_CERT = '-----BEGIN CERTIFICATE-----\ntest\n-----END CERTIFICATE-----';
    process.env.TUNNEL_CLIENT_CERT = '-----BEGIN CERTIFICATE-----\nclient\n-----END CERTIFICATE-----';
    process.env.TUNNEL_CLIENT_KEY = '-----BEGIN PRIVATE KEY-----\nkey\n-----END PRIVATE KEY-----';

    const cfg = loadConfig();
    expect(cfg!.caCert).toBeDefined();
    expect(cfg!.clientCert).toBeDefined();
    expect(cfg!.clientKey).toBeDefined();
  });

  it('reads optional TLS values from files when *_FILE vars are set', () => {
    const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'tunnel-agent-config-'));
    const caPath = path.join(tempDir, 'ca.pem');
    const certPath = path.join(tempDir, 'client.pem');
    const keyPath = path.join(tempDir, 'client.key');

    // Controlled temp files created by the test harness.
    // eslint-disable-next-line security/detect-non-literal-fs-filename
    fs.writeFileSync(caPath, 'ca-pem\n');
    // eslint-disable-next-line security/detect-non-literal-fs-filename
    fs.writeFileSync(certPath, 'client-pem\n');
    // eslint-disable-next-line security/detect-non-literal-fs-filename
    fs.writeFileSync(keyPath, 'key-pem\n');

    process.env.TUNNEL_SERVER_URL = 'wss://example.com/tunnel';
    process.env.TUNNEL_TOKEN = 'tok';
    process.env.TUNNEL_GATEWAY_ID = 'gw-1';
    process.env.TUNNEL_LOCAL_PORT = '4822';
    process.env.TUNNEL_CA_CERT_FILE = caPath;
    process.env.TUNNEL_CLIENT_CERT_FILE = certPath;
    process.env.TUNNEL_CLIENT_KEY_FILE = keyPath;

    const cfg = loadConfig();
    expect(cfg!.caCert).toBe('ca-pem');
    expect(cfg!.clientCert).toBe('client-pem');
    expect(cfg!.clientKey).toBe('key-pem');

    fs.rmSync(tempDir, { recursive: true, force: true });
  });
});

File: gateways/tunnel-agent/config_test.go
package main

import (
	"os"
	"path/filepath"
	"testing"
	"time"
)

func TestLoadConfigFromEnvDormant(t *testing.T) {
	clearTunnelEnv(t)
	t.Setenv("TUNNEL_LOCAL_PORT", "4822")

	cfg, dormant, err := LoadConfigFromEnv("test-version")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !dormant {
		t.Fatal("expected dormant mode")
	}
	if cfg != nil {
		t.Fatalf("cfg = %#v, want nil", cfg)
	}
}

func TestLoadConfigFromEnvRequiredAndDefaults(t *testing.T) {
	clearTunnelEnv(t)
	t.Setenv("TUNNEL_SERVER_URL", "https://arsenale.example.com")
	t.Setenv("TUNNEL_TOKEN", "tok")
	t.Setenv("TUNNEL_GATEWAY_ID", "gw")
	t.Setenv("TUNNEL_LOCAL_PORT", "4822")

	cfg, dormant, err := LoadConfigFromEnv("test-version")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if dormant {
		t.Fatal("did not expect dormant mode")
	}
	if cfg.ServerURL != "wss://arsenale.example.com/api/tunnel/connect" {
		t.Fatalf("ServerURL = %q", cfg.ServerURL)
	}
	if cfg.AgentVersion != "test-version" {
		t.Fatalf("AgentVersion = %q", cfg.AgentVersion)
	}
	if cfg.LocalServiceHost != "127.0.0.1" || cfg.LocalServicePort != 4822 {
		t.Fatalf("local service = %s:%d", cfg.LocalServiceHost, cfg.LocalServicePort)
	}
	if cfg.PingInterval != 15*time.Second || cfg.ReconnectInitial != time.Second || cfg.ReconnectMax != time.Minute {
		t.Fatalf("unexpected timing defaults: %#v", cfg)
	}
}

func TestLoadConfigFromEnvRejectsPartialConfig(t *testing.T) {
	clearTunnelEnv(t)
	t.Setenv("TUNNEL_SERVER_URL", "wss://example.com/tunnel")

	_, dormant, err := LoadConfigFromEnv("test-version")
	if dormant {
		t.Fatal("partial config should not be dormant")
	}
	if err == nil {
		t.Fatal("expected error")
	}
}

func TestLoadConfigFromEnvReadsPEMFilesAndInlineWins(t *testing.T) {
	clearTunnelEnv(t)
	dir := t.TempDir()
	caPath := filepath.Join(dir, "ca.pem")
	certPath := filepath.Join(dir, "client.pem")
	keyPath := filepath.Join(dir, "client.key")
	mustWrite(t, caPath, "file-ca\n")
	mustWrite(t, certPath, "file-cert\n")
	mustWrite(t, keyPath, "file-key\n")

	t.Setenv("TUNNEL_SERVER_URL", "wss://example.com/tunnel")
	t.Setenv("TUNNEL_TOKEN", "tok")
	t.Setenv("TUNNEL_GATEWAY_ID", "gw")
	t.Setenv("TUNNEL_LOCAL_PORT", "2222")
	t.Setenv("TUNNEL_CA_CERT", "inline-ca")
	t.Setenv("TUNNEL_CA_CERT_FILE", caPath)
	t.Setenv("TUNNEL_CLIENT_CERT_FILE", certPath)
	t.Setenv("TUNNEL_CLIENT_KEY_FILE", keyPath)

	cfg, _, err := LoadConfigFromEnv("test-version")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if cfg.CACert != "inline-ca" || cfg.ClientCert != "file-cert" || cfg.ClientKey != "file-key" {
		t.Fatalf("unexpected PEM values: %#v", cfg)
	}
}

func TestNormalizeTunnelServerURL(t *testing.T) {
	tests := map[string]string{
		"arsenale.example.com":                            "wss://arsenale.example.com/api/tunnel/connect",
		"https://arsenale.example.com":                    "wss://arsenale.example.com/api/tunnel/connect",
		"http://arsenale.example.com/base/":               "ws://arsenale.example.com/base/api/tunnel/connect",
		"wss://arsenale.example.com/custom":               "wss://arsenale.example.com/custom",
		"wss://arsenale.example.com/":                     "wss://arsenale.example.com/api/tunnel/connect",
		"https://arsenale.example.com/api/tunnel/connect": "wss://arsenale.example.com/api/tunnel/connect",
	}
	for raw, want := range tests {
		t.Run(raw, func(t *testing.T) {
			if got := normalizeTunnelServerURL(raw); got != want {
				t.Fatalf("normalizeTunnelServerURL(%q) = %q, want %q", raw, got, want)
			}
		})
	}
}

func clearTunnelEnv(t *testing.T) {
	t.Helper()
	for _, key := range []string{
		"TUNNEL_SERVER_URL", "TUNNEL_TOKEN", "TUNNEL_GATEWAY_ID", "TUNNEL_LOCAL_PORT",
		"TUNNEL_LOCAL_HOST", "TUNNEL_CA_CERT", "TUNNEL_CA_CERT_FILE",
		"TUNNEL_CLIENT_CERT", "TUNNEL_CLIENT_CERT_FILE", "TUNNEL_CLIENT_KEY",
		"TUNNEL_CLIENT_KEY_FILE", "TUNNEL_AGENT_VERSION", "TUNNEL_PING_INTERVAL_MS",
		"TUNNEL_RECONNECT_INITIAL_MS", "TUNNEL_RECONNECT_MAX_MS",
	} {
		key := key
		oldValue, hadValue := os.LookupEnv(key)
		os.Unsetenv(key)
		t.Cleanup(func() {
			if hadValue {
				_ = os.Setenv(key, oldValue)
				return
			}
			_ = os.Unsetenv(key)
		})
	}
}

func mustWrite(t *testing.T, path, contents string) {
	t.Helper()
	if err := os.WriteFile(path, []byte(contents), 0o600); err != nil {
		t.Fatalf("write %s: %v", path, err)
	}
}
